lib/pure/options.nim

```diff
@@ -176,21 +310,239 @@ proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+      else:
+        noneCase = optionCase[1]
+    of nnkCommand:
+      if $optionCase[0] != "some":
+        if $optionCase[0] != "none":
+          error "Only \"none\" and \"some\" are allowed as case labels",
+            optionCase[0]
+        else:
+          error "Only \"some\" is allowed to have arguments", optionCase[0]
+      else:
+        if optionCase[1].kind != nnkInfix:
+          if elseCase.body != nil:
+            error "Only one \"some\" case without a check is allowed, " &
+              "previously defined \"some\" case at: " & lineInfo(elseCase.body),
+              optionCase[0]
+          else:
+            if $optionCase[1] != "_":
+              elseCase = (ident: optionCase[1], body: optionCase[2])
+            else:
+              elseCase = (ident: nil, body: optionCase[2])
+        else:
+          if $optionCase[1][0] != "of":
+            error "Invalid separator, only \"of\" allowed", optionCase[1][0]
+          else:
+            if $optionCase[1][1] != "_":
+              cases.add (ident: optionCase[1][1], check: optionCase[1][2],
+                body: optionCase[2])
+            else:
+              cases.add (ident: nil, check: optionCase[1][2],
+                body: optionCase[2])
+    else:
+      error "Unrecognized structure of cases", optionCase
+  if noneCase == nil:
+    error "Must have a \"none\" case"
+  if elseCase.body == nil and cases.len == 0:
+    error "Must have a \"some\" case"
+  var
+    optionSym = genSym(nskLet)
+    value = nnkDotExpr.newTree(optionSym, newIdentNode("val"))
+  var caseStmt = nnkCaseStmt.newTree(
+    value
+  )
+  for c in cases:
+    var ofBranch = nnkOfBranch.newTree(
+      c.check,
+      newStmtList()
+    )
+    if c.ident != nil:
+      ofBranch[1].add newLetStmt(c.ident, value)
+    ofBranch[1].add c.body
+    caseStmt.add ofBranch
+  if elseCase.body != nil:
+    if elseCase.ident != nil:
+      caseStmt.add nnkElse.newTree(newStmtList(
+        newLetStmt(elseCase.ident, value),
+        elseCase.body
+      ))
+    else:
+      caseStmt.add nnkElse.newTree(elseCase.body)
+  result = quote do:
+    (proc (): auto =
+      let `optionSym` = `option`
+      if `optionSym`.isSome:
+        `caseStmt`
+      else:
+        `noneCase`
+    )()
+  #echo result.repr
+
+proc optCmp*[T](self: Option[T], value: Option[T],
+  cmp: proc (val1, val2: T): bool): Option[T] =
+  ## Comparison operation that return a some iff ``self`` is a some and the
+  ## value of ``cmp`` applied to ``self`` and ``value`` returns true. The
+  ## returned some has the value of ``self``.
+  if self.isSome and value.isSome and cmp(self.val, value.val):
+    return self
+
+template `==`*[T](self: Option[T], value: Option[T]): Option[T] =
+  ## Wrapper for optCmp with ``==`` as the comparator
+  optCmp(self, value, proc (val1, val2: T): bool = val1 == val2)
+
+template `not`*[T](self: Option[T]): Option[T] =
+  ## Not always returns a ``none`` as a ``none`` can't implicitly get a value
```

Comment:
WTF, what are you thinking here?

Reply:
It started as a workaround in that `x != y` get's rewritten to `not x == y` which breaks the logic of maybes. But when you think about it what would the `not` of a `none` be? You can't create a `some` out of it as that would imply it has a value..

Reply:
how about this:
```nim
template `not`*[T](self: Option[T]): bool = solf.isNone
```

Reply:
i think that entire idea of `not(Option[T])` smells bad. If you go this way, please also redefine and/or. And move it into separate module. Separate of stdlib. Main options.nim should return "==" result as bool, and avoid extending standard language functions with new meanings. Such level of changes to the language should be done after thorough discussion rather than on the spot.